### ISSUE TITLE
AGCM topo_* files and make_bcs

### ISSUE THREAD

**Issue by gmao-rreichle** [GEOS-ESM/GEOSgcm_GridComp#523]
This issue is a follow-up from #517 and is beyond the scope of #517.

1. Which package creates the AGCM's topo_* files?
2. Why is the generation of the topo_* files is separate from make_bcs?
3. Shouldn't the two sets of files be created as part of the same package and (ideally) use the same topo source data?

In the past, I pointed out potential inconsistencies between the topo_* files and the topography data underlying the Catchment parameter files, but I'm not aware of any follow-up. 

Something for the modeling group to keep in mind.

_Originally posted by @gmao-rreichle in https://github.com/GEOS-ESM/GEOSgcm_GridComp/issues/517#issuecomment-1015875367_

**Comment by bena-nasa** [comment]
@gmao-rreichle 
Here are are the answers to your questions (this is going to be long)
1. Somehow I've over the years become the person creates the topography files, probably because at one point Bill asked me and well I've just been doing it since then. There is not one package that creates these. Here are the details.
The topography is created from the 30" GTMED data with several correction we have to make which are the following:
I overlay SRTM 30" data over the antarctic penninsula as GMTED just has this wrong.
Both Lake Superior and the Caspian sea have issue (they are not a constant level but rather a hole ...) I have to fill those in to the correct level. NOAA/UFS also uses this as their starting point and also independently found the same problems they had to fix
Once those are fixed we have to create the topography for each grid resolution as well as the subgrid variances. For that have used this package:
https://github.com/NCAR/Topo
You provide a starting topography and scrip files describing your target grid and it generates the target grid topography as well as the two orographic variances relative to a c3000 cubed-sphere grid (this is configurable).
Once we have that I then have to smooth the topography. For what we currently use I simply ran these through a hacked version of the topography generation utility in FV3 provided by Bill to do the smoothing.
Once you have this I then regenerate the variances with the smoothed grid level topography.

Just last month Bill asked me to regenerate the topography but use the same smoothing program that is currently used by UFS. It took some work (I have to essentially build UFS on discover, which is doable if you can figure out how to get the right software stack built with their hpc tools, get grid description files in the format they want, gridspec rather than scrip, and then do data massaging to get it into a form their smoothing program wants) but I did get that working so I now have those for testing. 

In addition I ended simply running their equivalent full orographic generation work flow from the GMTED data as that was the easiest way to get the gridspec files I need and also produces the extra orography needed to use the NOAA gwd scheme.

In any case I have this procedure for automated for both the old and new smoothing.

2. No reason other than that it is quite involved as you can see. We would have to bring in that NCAR topography software, then bring in the appropriate smoothing software into GEOS so if we want to use the UFS smoothing code in GEOS we'd have to get that individual code building in our system. If we want to do this I don't see any technical reason, it just would take time, probably a good amount to get right and maintainable.

3. I don't know what to say about this point as what make_bcs does is really a black box for me other than some of the tile file generation code. Someone who understands what that is doing would need to answer that part.

**Comment by gmao-rreichle** [comment]
Many thanks, @bena-nasa, for the comprehensive update!  Very helpful to get a sense of what's involved and what the data sources are.  I think we can put item 1. aside. 

Re. item 2:  If it's too complicated to incorporate all the external packages into GEOS, perhaps the home-made pieces and instructions could be added to the GEOSgcm_GridComp under [..]/GEOSsurface_GridComp/Utils/Topo, that is, next to ./Raster (make_bcs) and ./mk_restarts directories.   This doesn't give us a ready-made package, and it's not integrated with make_bcs, but at least the information and scripts are in a discoverable and version-controlled location.

Re. item 3: Topography data is needed to generate the land bcs files.  The "old" ("Icarus") bcs use the outdated, coarse HYDRO1k data, the "new" ("Icarus-NLv3") bcs use SRTM (not sure about the exact flavor, but probably SRTM30) backfilled with other data sources at high lats (where SRTM has no-data-values).  
The make_bcs package includes elevation info for each land tile (in the file catchment.def, I think), but that is just for information and postprocessing and is not used in the model.  
Rather, the topography data was used in the delineation of the (Pfafstetter) catchments, which are hydrologically connected to permit streamflow routing.  This preprocessing of the topography data into catchments is a one-off process that was done by an outside collaborator and is **not** part of make_bcs.  
make_bcs only processes the ancillary catchments into tiles based on the selected atmospheric grid and resolution.
I don't know GMTED and how it might differ from the SRTM (+backfill) data used for make_bcs.  I suppose it's ok to use different topo inputs for make_bcs and the creation of the topo_* files for now.  Given that it seems to be a big effort to even create a self-contained package that creates the topo_* files within GEOS, and given that the preprocessing of the topo data into hydrological catchments is a one-off process, it seems that unifying the topo input data sources is out of reach. 

In summary, maybe pursue item 2.   But that's not my decision. 

additional cc: @rdkoster

**Comment by gmao-rreichle** [comment]
PS: Just for reference, here's an old discussion about unifying bcs scripts for the AGCM and the coupled model (not really related to topo files):  https://github.com/GEOS-ESM/GEOSgcm_App/discussions/193

**Comment by bena-nasa** [comment]
@gmao-rreichle 
I'd happy to see what can be captured as time permits at least as far as the higher level scripting goes and documenting. I think it would be an effort to create a self contained package for the topo generation, not impossible but time consuming. If someone had unfettered time to focus on this for a while it would be doable. There are just a lot of details to get everything to the final form GEOS expects but I imagine the same applies to make_bcs and what it does. I'll see what I can in the meantime.

**Comment by sanAkel** [comment]
@gmao-rreichle Thanks for starting this issue and @bena-nasa thanks for the details - I haven't read it!

However, I want to make sure of one thing right away! A quick "scan" tells me that all of this has no **ocean** topography. Is that correct?? In other words, GEOS does not care about ocean depths and that should include coastline. 

If that is correct, we bring our topography and somehow have to connect with what all is done (above)? I say/ask this way because the topography that we work with: https://www.gebco.net/ has data over land as well https://www.gebco.net/data_and_products/gridded_bathymetry_data/ at 15 arc-seconds (FYI).

**Comment by gmao-rreichle** [comment]
@sanAkel, thanks for pointing out the fact that the ocean also has its own topography...  That's what you get from a landlubber such as myself ;)
I'm not sure whether the s/w for ocean (bottom) topography would be most appropriately placed into the Surface GC folder (ie, next to ./Raster, ./mk_restarts, and the "atm" topo_* stuff).  Definitely something else to consider, no matter where things end up.

**Comment by sanAkel** [comment]
@gmao-rreichle, I would say something _similar_ about myself (in the ocean)! 😄 

>I'm not sure whether the s/w for ocean (bottom) topography would be most appropriately placed into the Surface GC folder (ie, next to ./Raster, ./mk_restarts, and the "atm" topo_* stuff).

I would suggest that unless there are _dependencies_ or reasons?? we should pull all such software (topo/grid generation) into a new repo of its own -- @tclune? I don't see point why it should all fall into the `GEOSgcm_GridComp/` Also the time it takes for PR approval in this repo is not _short_ and what kind of tests- unity/sanity exist as of now in `GEOSgcm_GridComp/` specifically to check for changes in such grids and topo- AFAIK none! Hence I see no dependency.

My preference, we pull out, re-org and yes, ocean should be part of that - we commit everything there!

(I will reach out to you via MS teams on a related topic.)

**Comment by gmao-rreichle** [comment]
@sanAkel:   For the land, it probably doesn't make sense to pull make_bcs out of the GEOSgcm_GridComp.  The land bcs generation is tightly integrated with the land model physics because the land parameters in the bcs are very specific to the Catchment model.  It's probably best to keep it the land bcs and the land model in the same repo.  But if anyone has a concrete solution for disentangling the bcs and the land model elements, I'm happy to entertain it.

**Comment by sanAkel** [comment]
> The land bcs generation is tightly integrated with the land model physics because the land parameters in the bcs are very specific to the Catchment model. 

I did not know that, in that case it makes sense to keep it all together.

However, that _design_ may lead to more entangled dependencies and not sure how complicated it would be to keep backward compatibility and portability. I guess there are choices that have been made, after weighing in pros/cons.

As for the ocean, it is all totally independent. Once we have a clear picture, my intention is to make it somehow hooked to the GEOS-ESM.

**Comment by tclune** [comment]
Apologies in advance if I've not quite gotten the full context - these are tools that I'm not very familiar with.   To my limited understanding this is software that is used to _generate_ data that is then ingested by runs of the model.  The "plan" (Arlindo's vision) is that such software should be placed in a subdirectory of the "data tank" alongside the data that it generates.     But lacking familiarity I don't know if this will make sense in all cases.   And then whether this placement is in a repository or just a secure (backed-up) directory is also up for grabs.   My plan had been to put the data into git (using an extension called git-annex for large binary data), but other than versioning these tools there was not much bang for the effort.   

If I'm way off base, please educate me.

**Comment by sanAkel** [comment]
@tclune, I don't think this plan:

> The "plan" (Arlindo's vision) is that such software should be placed in a subdirectory of the "data tank" alongside the data that it generates
...
...
 My plan had been to put the data into git (using an extension called git-annex for large binary data), but other than versioning these tools there was not much bang for the effort.
> 

would work since @gmao-rreichle said [above](https://github.com/GEOS-ESM/GEOSgcm_GridComp/issues/523#issuecomment-1018082745) that:

>The land bcs generation is tightly integrated with the land model physics because the land parameters in the bcs are very specific to the Catchment model. 

May be @amdasilva would be able to help us ...

**Comment by gmao-rreichle** [comment]
@tclune, thanks for the insights.  Here are a couple of responses:

> To my limited understanding this is software that is used to generate data that is then ingested by runs of the model. 

That's correct. 

> The "plan" (Arlindo's vision) is that such software should be placed in a subdirectory of the "data tank" alongside the data that it generates. 

That sounds sensible, until you get to the land parameter generation, which shares a bunch of s/w with the model.   It can probably be disentangled and appropriate libraries could be created, but we'd probably still end up with very model-specific stuff in an external repository, which would make it only harder to keep the bcs s/w in sync with the model. 

> And then whether this placement is in a repository or just a secure (backed-up) directory is also up for grabs. My plan had been to put the data into git (using an extension called git-annex for large binary data), but other than versioning these tools there was not much bang for the effort.

I can't comment on where exactly to place the data.  At the moment, we don't even have a (nobackup) project directory for them, or anyone in charge of maintaining the data. 

As the discussion above shows, some of the s/w that generates ancillary datasets does not even seem to be under version control yet.  There's a long way to go here...
